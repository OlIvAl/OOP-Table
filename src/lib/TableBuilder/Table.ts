import { IColumn, ITable } from './interfaces';

export class Table<R> implements ITable<R> {
  columns: Record<keyof R, IColumn<R>> | null = null;
  rows: R[] = [];
  footerRow: Record<keyof R, string> | null = null;
}
