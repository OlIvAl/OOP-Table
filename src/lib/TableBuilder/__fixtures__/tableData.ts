import { IColumn, IModelsToRowsConverter } from '../interfaces';
// import * as faker from 'faker';

export interface ITestModel {
  value1: number;
  value2: string;
  value3: {
    subValue1: number;
    subValue2: string;
  };
}
export interface ITestRow {
  col1: string;
  col2: string;
  col3: string;
}

export const modelsToRowsConverter: IModelsToRowsConverter<ITestModel, ITestRow> = model => {
  return {
    col1: model.value1.toString(),
    col2: model.value2,
    col3: `${model.value3.subValue1} ${model.value3.subValue2}`
  };
};

export const columnKeys: (keyof ITestRow)[] = ['col1', 'col2', 'col3'];

export const columns: Record<keyof ITestRow, IColumn<ITestRow>> = {
  col1: {
    accessor: columnKeys[0],
    headerClassName: columnKeys[0]
  },
  col2: {
    accessor: columnKeys[1],
    headerClassName: columnKeys[1]
  },
  col3: {
    accessor: columnKeys[2],
    headerClassName: columnKeys[2]
  }
};

export const getFakeModels = () => [
  { value1: 7, value2: 'ADP', value3: { subValue1: 76, subValue2: 'compelling' } },
  { value1: 47, value2: 'evolve', value3: { subValue1: 47, subValue2: 'array' } },
  { value1: 93, value2: 'Sleek', value3: { subValue1: 100, subValue2: 'grow' } }
];

export const getFooterRow = () => ({ col1: '65', col2: '57', col3: '' });
