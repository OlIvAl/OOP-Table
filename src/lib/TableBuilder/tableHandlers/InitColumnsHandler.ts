import { AbstractHandler } from './AbstractHandler';
import { IColumn, ITable } from '../interfaces';

export class InitColumnsHandler<M, R> extends AbstractHandler<M, R> {
  /**
   * ключи колонок
   */
  private columnKeys: (keyof R)[];

  constructor(columnKeys: (keyof R)[]) {
    super();

    this.columnKeys = columnKeys;

    this.makeColumn = this.makeColumn.bind(this);
  }

  /**
   * Инициализирует поле колонок таблицы
   * @param table объект таблицы
   * @returns модифицированная таблица
   */
  protected transform(table: ITable<R>): ITable<R> {
    table.columns = this.columnKeys.reduce<Record<keyof R, IColumn<R>>>(
      (record, key) => ({ ...record, [key]: this.makeColumn(key) }),
      {} as Record<keyof R, IColumn<R>>
    );
    return table;
  }

  /**
   * Преобразует ключ колонки в обЪект колонки
   * @param columnKey ключ колонки
   * @returns обЪект колонки
   */
  private makeColumn(columnKey: keyof R): IColumn<R> {
    return {
      accessor: columnKey,
      headerClassName: columnKey
    };
  }
}
