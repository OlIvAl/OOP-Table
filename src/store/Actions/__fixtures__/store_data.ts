
// import * as faker from 'faker';

import { RootStore } from '../../RootStore';
import { IStoreWithCurrentModel } from '../../Abstract/interfaces';
import { IFinanceActionModel } from '../interfaces';
import { FinanceActionsStore } from '../FinanceActionsStore';
import DataLayer from '../../../DataLayer';
import { FinanceActionModel } from '../FinanceActionModel';
import { IFinanceActionData } from '../../../DataLayer/modules/actions/interfaces';

const rootStore = new RootStore();

export const getActionsStoreFromDataLayerMocks = async (): Promise<IStoreWithCurrentModel<IFinanceActionModel>> => {
  const store = new FinanceActionsStore(rootStore);
  const response = await DataLayer.getBalanceStateData({});

  store.models = response.actions.map<IFinanceActionModel>((data: IFinanceActionData) => {
    const model = new FinanceActionModel(store, data.action_id);

    model.action_id = data.action_id;
    model.campaign_id = data.campaign_id;
    model.campaign_name = data.campaign_name;
    model.marker = data.marker;
    model.source = data.source;
    model.action_state = data.action_state;
    model.created_at = data.created_at;
    model.updated_at = data.updated_at;
    model.price = data.price;
    model.description = data.description;
    model.profit_diff = data.profit_diff;
    model.total_profit = data.total_profit;

    return model;
  });

  return store;
};

/*export const getStoreFromFaker = (): IFinanceActionsStore => {
  const store = new FinanceActionsStore(rootStore);

  const models: IFinanceActionModel[] = [];

  for (let i = 0; i < 50; i++) {
    const id = (faker.random.uuid() as unknown) as number;
    models[i] = new FinanceActionModel(store, id);

    models[i].action_id = id;
    models[i].campaign_id = faker.random.number(1000);
    models[i].campaign_name = faker.company.companyName();
    models[i].marker = faker.random.uuid();
    models[i].source = faker.random.word();
    models[i].action_state = faker.helpers.randomize(['paid', 'cancelled', 'processing']);
    models[i].created_at = faker.date.past().toDateString();
    models[i].updated_at = faker.date.past().toDateString();
    models[i].price = faker.helpers.randomize([faker.random.number(1000000), null]);
    models[i].description = faker.lorem.sentences(3);
    models[i].profit_diff = faker.random.number(100);
    models[i].total_profit = faker.random.number(1000);
  }

  store.models = models;

  return store;
};*/
