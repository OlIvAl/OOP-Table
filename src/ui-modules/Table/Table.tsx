import React from 'react';
import { TableOptions, useTable } from 'react-table';

interface IProps extends TableOptions<any> {
  loading: boolean
}

export const Table: React.FC<IProps> = ({ columns, data , loading}): JSX.Element => {
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
  } = useTable({
    columns,
    data,
  });

  if( loading ) {
    return <div>Loading...</div>;
  }

  return (
    <table {...getTableProps()}>
      <thead>
      {headerGroups.map(headerGroup => (
        <tr {...headerGroup.getHeaderGroupProps()}>
          {headerGroup.headers.map(column => (
            <th {...column.getHeaderProps()}>{column.render('Header')}</th>
          ))}
        </tr>
      ))}
      </thead>
      <tbody {...getTableBodyProps()}>
      {rows.map(
        (row) => {
          prepareRow(row);
          return (
            <tr {...row.getRowProps()}>
              {row.cells.map(cell => {
                return <td {...cell.getCellProps()}>{cell.render('Cell')}</td>
              })}
            </tr>
          )}
      )}
      </tbody>
    </table>
  );
};