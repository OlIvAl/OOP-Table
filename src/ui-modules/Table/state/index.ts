import { action, computed, flow, IReactionDisposer, reaction } from 'mobx';
import { BalanceRowDecorator } from './decorators/BalanceRowDecorator';
import { BalanceTotalRowDecorator } from './decorators/BalanceTotalRowDecorator';
import { ITable } from '../../../lib/TableBuilder/interfaces';
import { IFinanceActionModel } from '../../../store/Actions/interfaces';
import { TableBuilder } from '../../../lib/TableBuilder/TableBuilder';
import { IBalanceParams, IFinanceActionData } from '../../../DataLayer/modules/actions/interfaces';
import { IStoreWithCurrentModel } from '../../../store/Abstract/interfaces';
import { IBalanceRow, IBalanceTotalRow, IFinanceActionsDataService } from './interfaces';
import { FinanceActionsStoreInitDecorator } from './decorators/FinanceActionsStoreInitDecorator';

/**
 * Объект состояния компонента баланса
 */
export class BalanceTableState {
  @computed get loading(): boolean {
    return this.getFinanceActionsService.loading;
  }
  @computed get balanceTableShown(): boolean {
    return !!this.financeActionsStore.models.length;
  }
  @computed get table(): ITable<IBalanceRow> | null {
    const models: IFinanceActionModel[] = this.financeActionsStore.models;
    if (!models.length) {
      return null;
    }

    const rows: IBalanceRow[] = models.map<IBalanceRow>(model => new BalanceRowDecorator(model));

    const balanceTableBuilder = new TableBuilder<IBalanceRow>();
    const totalRow: IBalanceTotalRow | null = this.getFinanceActionsService.total
      ? new BalanceTotalRowDecorator(this.getFinanceActionsService.total, this.currency)
      : null;

    balanceTableBuilder.setColumnKeys([
      'source',
      'created_at',
      'updated_at',
      'description',
      'action_state',
      'price',
      'profit_diff',
      'total_profit'
    ]);
    balanceTableBuilder.setModels(rows);
    balanceTableBuilder.setFooter(totalRow);

    return balanceTableBuilder.getTable();
  }

  protected financeActionsStore: IStoreWithCurrentModel<IFinanceActionModel>;
  protected getFinanceActionsService: IFinanceActionsDataService;
  protected financeActionsDataHandler: IReactionDisposer;

  protected get currency(): string {
    return 'rub';
  }

  constructor(
    financeActionsStore: IStoreWithCurrentModel<IFinanceActionModel>,
    getFinanceActionsService: IFinanceActionsDataService
  ) {
    this.financeActionsStore = financeActionsStore;
    this.getFinanceActionsService = getFinanceActionsService;

    this.financeActionsDataHandler = reaction<IFinanceActionData[]>(
      () => this.getFinanceActionsService.actions,
      (actions): void => {
        new FinanceActionsStoreInitDecorator(this.financeActionsStore).init(actions);
      }
    );
  }

  @action.bound async init(): Promise<void> {
    const params: IBalanceParams = {};

    await flow(this.getFinanceActionsService.init)(params);
  }

  /**
   * Метод получения списка букингов по передаваемым параметрам
   * @param params Объект с данными для запроса букингов
   */
  @action.bound async getFinanceActions(params: IBalanceParams): Promise<void> {
    await flow(this.getFinanceActionsService.init)(params);
  }

  /**
   * Сбрасывает хранилище
   */
  @action.bound dispose(): void {
    this.financeActionsStore.resetStore();
    this.financeActionsDataHandler();
  }
}
