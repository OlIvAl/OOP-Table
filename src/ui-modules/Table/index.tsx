import React from 'react';
import { Table } from './Table';
import { observable } from 'mobx';
import { RootStore } from '../../store/RootStore';
import { StoreContext } from '../../index';
import { BalanceTableState } from './state';
import { observer } from 'mobx-react-lite';
import { FinanceActionsDataService } from './services/FinanceActionsDataService';

const BalanceTable: React.FC = (): JSX.Element | null => {
  const getFinanceActionsService = new FinanceActionsDataService();
  const { financeActionsStore } = React.useContext<RootStore>(StoreContext);
  const [balanceTableState] = React.useState<BalanceTableState>(() =>
    observable(new BalanceTableState(financeActionsStore, getFinanceActionsService))
  );

  React.useEffect(() => {
    balanceTableState.init();

    return balanceTableState.dispose;
  }, [balanceTableState]);

  const table = balanceTableState.table;

  if (!table || !table.columns) {
    return null;
  }

  const columns = [
    {
      Header: () => 'Source',
      headerClassName: table.columns.source.headerClassName,
      accessor: table.columns.source.accessor
    },
    {
      Header: () => 'Created at',
      accessor: table.columns.created_at.accessor,
      headerClassName: table.columns.created_at.headerClassName
    },
    {
      Header: () => 'Updated at',
      accessor: table.columns.updated_at.accessor,
      headerClassName: table.columns.updated_at.headerClassName
    },
    {
      Header: () => 'Description',
      headerClassName: table.columns.description.headerClassName,
      accessor: table.columns.description.accessor
    },
    {
      Header: () => 'Price',
      headerClassName: table.columns.price.headerClassName,
      accessor: table.columns.price.accessor
    },
    {
      Header: () => 'Total profit',
      headerClassName: table.columns.total_profit.headerClassName,
      accessor: table.columns.total_profit.accessor
    }
  ];

  return (
    <section className="balance_actions_table">
      <Table
        data={table.rows}
        columns={columns}
        loading={balanceTableState.loading}
      />
    </section>
  );
};

export default observer(BalanceTable);