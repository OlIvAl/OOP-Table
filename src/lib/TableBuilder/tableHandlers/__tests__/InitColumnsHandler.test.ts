import { columnKeys, columns, ITestModel, ITestRow } from '../../__fixtures__/tableData';
import { InitColumnsHandler } from '../InitColumnsHandler';
import { Table } from '../../Table';

let initColumnsHandler: InitColumnsHandler<ITestModel, ITestRow>;
const table = new Table<ITestRow>();

beforeEach(() => {
  initColumnsHandler = new InitColumnsHandler<ITestModel, ITestRow>(columnKeys);
});

describe('InitColumnsHandler', () => {
  describe('constructor', () => {
    it('init columns handler with valid object', () => {
      expect(initColumnsHandler).toMatchSnapshot();
    });
  });
  describe('handle', () => {
    it('correctly init columns', () => {
      initColumnsHandler.handle(table);

      expect(table.columns).toEqual(columns);
    });
  });
});
