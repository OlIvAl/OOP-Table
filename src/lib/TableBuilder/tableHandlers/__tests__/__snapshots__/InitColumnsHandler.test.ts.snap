// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitColumnsHandler constructor init columns handler with valid object 1`] = `
InitColumnsHandler {
  "columnKeys": Array [
    "col1",
    "col2",
    "col3",
  ],
  "handle": [Function],
  "makeColumn": [Function],
  "nextHandler": null,
  "setNext": [Function],
  "transform": [Function],
}
`;
